nixos/modules/services/misc/matrix-appservice-discord.nix

```diff
@@ -0,0 +1,118 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  dataDir = "/var/lib/matrix-appservice-discord";
+in {
+  options = {
+    services.matrix-appservice-discord = {
+      enable = mkEnableOption "a bridge between Matrix and Discord";
+
+      port = mkOption {
+        type = types.int;
+        default = 9005; # from https://github.com/Half-Shot/matrix-appservice-discord/blob/master/package.json#L11
+        description = ''
+          Port number on which the bridge should listen for internal communication with the Matrix homeserver.
+        '';
+      };
+
+      config = mkOption rec {
+        type = types.attrs;
+        apply = recursiveUpdate default;
+        default = {
+          database = {
+            filename = "${dataDir}/discord.db";
+          };
+        };
+        example = literalExample ''
+          {
+            bridge = {
+              domain = "public-domain.tld";
+              homeserverUrl = "http://public-domain.tld:8008";
+            };
+          }
+        '';
+        description = ''
+          <filename>config.yaml</filename> configuration as a Nix attribute set.
+
+          Configuration options should match those described in
+          <link xlink:href="https://github.com/Half-Shot/matrix-appservice-discord/blob/master/config/config.sample.yaml">
+          config.sample.yaml</link>.
+
+          <option>config.bridge.domain</option> and <option>config.bridge.homeserverUrl</option>
+          should be set to match the public host name of the Matrix homeserver for webhooks and avatars to work.
+
+          Secret tokens should be specified using <option>environmentFile</option>
+          instead of this world-readable attribute set.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = types.nullOr types.path;
```

Comment:
<del>Is there any reason to not use an attribute set to represent the environment variables? This makes it easier to edit the environment from multiple modules.</del>

EDIT: This might be a bad idea actually as it should be possible to have environment files outside of the store (to avoid that secrets are world-readable on a machine). However you should do `either (path str)` to avoid that the file will be copied to the store (which is the case when using Nix paths).

Reply:
Should I set it to `with types; nullOr (either (path str))`?

Reply:
Missed the `nullIOr` part while reading the type declaration, but this seems reasonable :+1:

Reply:
Actually, I'm not sure whether I understand correctly.
Is `either (path str)` preventing the copy into the store, even if the user specifies a `path`?
If not, maybe should we simply put `type = nullOr string` instead.

Reply:
Apparently, a `path` is copied in the nix store when it's used in `${}`.
This is not the case when passed to `EnvironmentFile` in `serviceConfig`.
For typing, I suggest that we leave it as `type = nullOr path`.

Reply:
You're right. The reason can be found in [`nixos/modules/system/boot/systemd-lib.nix`](https://github.com/NixOS/nixpkgs/blob/70de27bbb4758038e5b87f725d964eefda4898e1/nixos/modules/system/boot/systemd-lib.nix#L107): each value from `serviceConfig` will be converted to a string using `toString` before substituting it with `${}` into the config string.

I didn't know about the `toString` step here, so it's perfectly fine to use `nullOr path` :+1: